protobuf_serialization.nim

```diff
@@ -0,0 +1,376 @@
+  let wireTy = wireType(tyByte)
+  if wireTy != Varint:
+    raise newException(UnexpectedTypeError, fmt"Not a varint at offset {offset}! Received a {wireTy}")
+
+proc checkType[T: SomeFixed](tyByte: byte, ty: typedesc[T], offset: int) {.inline.} =
+  let wireTy = wireType(tyByte)
+  if wireTy notin {Fixed32, Fixed64}:
+    raise newException(UnexpectedTypeError, fmt"Not a fixed32 or fixed64 at offset {offset}! Received a {wireTy}")
+
+proc checkType[T: SomeLengthDelimited](tyByte: byte, ty: typedesc[T], offset: int) {.inline.} =
+  let wireTy = wireType(tyByte)
+  if wireTy != LengthDelimited:
+    raise newException(UnexpectedTypeError, fmt"Not a length delimited value at offset {offset}! Received a {wireTy}")
+
+proc checkType[T: object](tyByte: byte, ty: typedesc[T], offset: int) {.inline.} =
+  let wireTy = wireType(tyByte)
+  if wireTy != LengthDelimited:
+    raise newException(UnexpectedTypeError, fmt"Not an object value at offset {offset}! Received a {wireTy}")
+
+proc get*[T: SomeLengthDelimited](
+  bytes: var seq[byte],
+  ty: typedesc[T],
+  outOffset: var int,
+  outBytesProcessed: var int,
+  numBytesToRead = none(int)
+): T {.inline.} =
+  var bytesRead = 0
+  let decodedSize = bytes.get(uint, outOffset, outBytesProcessed, numBytesToRead)
+  let length = decodedSize.int
+
+  when T is string:
+    result = newString(length)
+    for i in outOffset ..< (outOffset + length):
+      result[i - outOffset] = bytes[i].chr
+  elif T is cstring:
+    result = cast[cstring](bytes[outOffset ..< (outOffset + length)])
+  else:
+    result.setLen(length)
+    for i in outOffset ..< (outOffset + length):
+      result[i - outOffset] = type(result[0])(bytes[i])
+
+  increaseBytesRead(length)
+
+proc decodeField*[T: SomeFixed | SomeVarint | SomeLengthDelimited](
+  bytes: var seq[byte],
+  ty: typedesc[T],
+  outOffset: var int,
+  outBytesProcessed: var int,
+  numBytesToRead = none(int)
+): ProtoField[T] {.inline.} =
+  var bytesRead = 0
+
+  checkType(bytes[outOffset], ty, outOffset)
+
+  result.index = fieldNumber(bytes[outOffset])
+  increaseBytesRead()
+
+  result.value = bytes.get(ty, outOffset, outBytesProcessed, numBytesToRead)
+
+proc decodeField*[T: object](
+  bytes: var seq[byte],
+  ty: typedesc[T],
+  outOffset: var int,
+  outBytesProcessed: var int,
+  numBytesToRead = none(int)
+): ProtoField[T] {.inline.}
+
+proc decodeField*[T: not AnyProtoType](
+  bytes: var seq[byte],
+  ty: typedesc[T],
+  outOffset: var int,
+  outBytesProcessed: var int,
+  numBytesToRead = none(int)
+): ProtoField[T] {.inline.} =
+
+  var bytesRead = 0
+
+  checkType(bytes[outOffset], seq[byte], outOffset)
+
+  result.index = fieldNumber(bytes[outOffset])
+  increaseBytesRead()
+
+  var value = bytes.get(seq[byte], outOffset, outBytesProcessed, numBytesToRead)
+  result.value = value.to(T)
+
+macro setField(obj: typed, fieldNum: int, offset: int, bytesProcessed: int, bytesToRead: Option[int], value: untyped): untyped =
```

Comment:
This won't be needed if you follow the pattern of using a `FieldReadersTable` as suggested above. If you deem it necessary nonetheless, I think it can be moved to nim-serialization. This would fit better the general division of responsibilities between the libraries. nim-serialization tries to host general-purpose procs and macros for analyzing the serialized types and generating code derived from their structure.